chain-evm/src/machine.rs

```diff
@@ -98,6 +112,73 @@ fn precompiles(fork: HardFork) -> Precompiles {
     }
 }
 
+impl<'runtime> VirtualMachine<'runtime> {
+    fn execute_transaction<F, T>(
+        &mut self,
+        gas_limit: u64,
+        delete_empty: bool,
+        f: F,
+    ) -> Result<(&AccountTrie, &LogsState, T), Error>
+    where
+        F: Fn(
+            &mut StackExecutor<
+                'runtime,
+                '_,
+                MemoryStackState<'_, 'runtime, VirtualMachine>,
+                Precompiles,
+            >,
+            u64,
+        ) -> (ExitReason, T),
+    {
+        let executable = |gas_limit| {
+            let metadata = StackSubstateMetadata::new(gas_limit, self.config);
+            let memory_stack_state = MemoryStackState::new(metadata, self);
+            let mut executor = StackExecutor::new_with_precompiles(
+                memory_stack_state,
+                self.config,
+                &self.precompiles,
+            );
+            (f(&mut executor, gas_limit), executor)
+        };
+
+        let ((exit_reason, val), executor) = executable(gas_limit);
+        match exit_reason {
+            ExitReason::Succeed(_) => {
+                // apply and return state
+                // apply changes to the state, this consumes the executor
+                let state = executor.into_state();
+                // Next, we consume the stack state and extract the values and logs
+                // used to modify the accounts trie in the VirtualMachine.
+                let (values, logs) = state.deconstruct();
+
+                self.apply(values, logs, delete_empty);
+                // exit_reason
+                Ok((&self.state, &self.logs, val))
+            }
+
+            // WARNING: such approach was followed from the https://github.com/paritytech/frontier/blob/4a77d59d4640e492ae0db43f891929b75876cf46/client/rpc/src/eth.rs#L1447
+            // need to clarify it
+
+            // If the transaction reverts, there are two possible cases,
+            // it can revert because the called contract feels that it does not have enough
+            // gas left to continue, or it can revert for another reason unrelated to gas.
+            ExitReason::Revert(err) => {
+                // If the user has provided a gas limit or a gas price, then we have executed
+                // with less block gas limit, so we must reexecute with block gas limit to
+                // know if the revert is due to a lack of gas or not.
+                let block_gas_limit = self.environment.block_gas_limit;
+                let ((exit_reason, _), _) = executable(block_gas_limit.as_u64());
+                match exit_reason {
+                    ExitReason::Succeed(_) => Err(Error::NotEnoughGas),
+                    _ => Err(Error::TransactionRevertError(err)),
+                }
```

Comment:
My judgement is that it is actually incorrect. The assumption here is that any transaction can fit into the `block_gas_limit`, while there is no such guarantee. Although I cannot immediately see what can be done here, probably nothing if the `evm` crate does not provide the separate out-of-gas error type.

Reply:
As I mentioned before I have followed to the https://github.com/paritytech/frontier/blob/4a77d59d4640e492ae0db43f891929b75876cf46/client/rpc/src/eth.rs#L1447, where considers the same situation. Actually we do not have such assumtion that any transaction can fit with the block_gas_limit, intention of this validation to precisly determine the reason of the rejection of this execution `let ((exit_reason, val), executor) = executable(gas_limit);`.